### ISSUE TITLE
[Feature Request] Sync Plex Labels to Jellyfin Tags

### ISSUE THREAD

**Issue by stquinn** [luigi311/JellyPlex-Watched#188]
**Is your feature request related to a problem? Please describe.**

I am looking to move from plex to jellyfin and I use Labels in Plex to whitelist shows and movies for my kids accounts.  I would like replicate this in Jellyfin.  With Jellyfin I can do the same with Tags.  I was wondering if it would be possible to add an option to move Plex Labels to Jellyfin as tags?

**Describe the solution you'd like**

Labels added to shows and movies in Plex added to jellyfin as tags

**Describe alternatives you've considered**

Doing it by hand - my libraries are not small and would be extremely time consuming..

**Additional context**

**Comment by lightmaster** [comment]
Would love to see this too, as I also use Plex labels to whitelist shows for my kids.

**Comment by luigi311** [comment]
Honestly i didnt even know that was a thing. Im also not sure if that would even fit with this tool but ill keep it in mind and see.

**Comment by lightmaster** [comment]
Yea, it's a simple and easy way to make sure your kids don't watch movies that aren't appropriate for them. Don't like the idea of splitting the library into "Kid's Movies" and "Movies" and "Kid's Shows" and "TV Shows", just seems to keep it simpler to have a single library for Movies and one for TV Shows, and use the built-in restrictions to limit what they can watch.

I know it doesn't match the scope of this tool 100%, but it doesn't seem that far out of it either and seems to still fall into the same idea of allowing syncing between Plex and Jellyfin. For the settings side of it, could be something simple like providing a matching list ("plexLabel": "jellyTag") and any content in plex with a matching plexLabel gets synced to Jellyfin with the jellyTag, and vice versa. It would definitely have to be a whitelist-type thing on the Jellyfin side cuz while Plex uses labels to restrict content and doesn't do anything else with them, Jellyfin uses tags for restrictions and does automatically populate them with something akin to genres.

**Comment by stquinn** [comment]
> Honestly i didnt even know that was a thing. Im also not sure if that would even fit with this tool but ill keep it in mind and see.

Yeah I guess if you consider the scope of this tool to sync the user modifiable state of media items from plex to jellyfin and vice versa, then labels/tags might fit into that idea.

Would be great, if it worked out. Thanks for considering it.

**Comment by stquinn** [comment]
@lightmaster in case this is still of interest here is a script that will sync Plex Labels to Jellyfin Tags.  It appears to work fine but I give absolutely no guarantees, so use at your own risk.  Make use of the --dry-run command line option.


`import argparse
import logging
from xml.etree import ElementTree
from plexapi.server import PlexServer
from jellyfin_apiclient_python import JellyfinClient

# ----------------- SETUP LOGGING ------------------
def setup_logging(to_stdout=False):
    formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
    root = logging.getLogger()
    root.setLevel(logging.INFO)
    for handler in root.handlers[:]:
        root.removeHandler(handler)
    if to_stdout:
        ch = logging.StreamHandler()
        ch.setFormatter(formatter)
        root.addHandler(ch)
    else:
        fh = logging.FileHandler("migration_log.txt")
        fh.setFormatter(formatter)
        root.addHandler(fh)

# ----------------- PLEX SECTION ------------------
def get_plex_items(plex_url, plex_token, match_by):
    plex = PlexServer(plex_url, plex_token)
    ids_to_labels = {}
    headers = {"X-Plex-Token": plex_token}

    for section in plex.library.sections():
        if section.TYPE not in ["movie", "show"]:
            continue
        for item in section.all():
            label_tags = [label.tag for label in getattr(item, 'labels', [])]
            if not label_tags:
                continue
            rating_key = item.ratingKey
            matched_id = None
            guid = item.guid or ""

            if f"{match_by}://" in guid:
                matched_id = guid.split(f"{match_by}://")[1].split("?")[0]
            else:
                try:
                    meta_url = f"{plex_url}/library/metadata/{rating_key}"
                    import requests
                    response = requests.get(meta_url, headers=headers, timeout=5)
                    response.raise_for_status()
                    root = ElementTree.fromstring(response.content)
                    for g in root.findall(".//Guid"):
                        gid = g.attrib.get("id", "")
                        if gid.startswith(f"{match_by}://"):
                            matched_id = gid.split(f"{match_by}://")[1]
                            break
                except Exception as e:
                    logging.warning(f"Metadata fetch failed for {item.title}: {e}")
                    continue

            if matched_id:
                ids_to_labels[matched_id] = {
                    "title": item.title,
                    "labels": label_tags
                }
                logging.info(f"Matched {item.title} -> {match_by.upper()} ID {matched_id}")

    return ids_to_labels

# ----------------- JELLYFIN SECTION ------------------
def jellyfin_connect(jellyfin_url, username, password):
    client = JellyfinClient()
    client.config.app("PlexToJellyfinSync", "1.0", "label-sync-client", "unique-id")
    client.config.data["auth.ssl"] = jellyfin_url.startswith("https")
    client.config.data["auth.serveraddress"] = jellyfin_url

    try:
        client.auth.connect_to_address(jellyfin_url)
        client.auth.login(jellyfin_url, username, password)
        return client
    except Exception as e:
        logging.error(f"❌ Failing server connection. ERROR msg: {e}")
        raise SystemExit(1)
    except Exception as e:
        logging.error(f"❌ Failing server connection. ERROR msg: {e}")
        raise SystemExit(1)

def find_jellyfin_item(client, match_by, external_id):
    try:
        results = client.jellyfin._get("Items", params={
            "Recursive": True,
            "IncludeItemTypes": "Movie,Series",
            "Fields": "ProviderIds",
            "api_key": client.config.data["auth.token"]
        })
        for item in results.get("Items", []):
            provider_ids = item.get("ProviderIds", {})
            if provider_ids.get(match_by.capitalize()) == external_id:
                item_type = item.get("Type")
                print(f"🧩 Found Jellyfin item match: {item['Name']} ({item_type}) ID={item['Id']}")
                if item_type not in ["Movie", "Series"]:
                    logging.warning(f"Skipping unsupported Jellyfin item type: {item_type} for {item['Name']}")
                    continue
                return item["Id"], item.get("Name")
    except Exception as e:
        logging.error(f"❌ Failed to find item in Jellyfin: {e}")
    return None, None

def apply_tags(client, item_id, new_tags, dry_run):
    try:
        metadata = client.jellyfin._get(f"Items/{item_id}")
    except Exception as e:
        logging.error(f"❌ Failed to fetch item metadata: {e}")
        return

    # Strip down metadata to only safe, updatable fields
    allowed_keys = [
        "Name", "OriginalTitle", "SortName", "ForcedSortName",
        "Tags", "Genres", "Studios", "ProviderIds",
        "PremiereDate", "ProductionYear", "OfficialRating",
        "Overview", "CommunityRating", "RunTimeTicks"
    ]
    update_payload = {k: v for k, v in metadata.items() if k in allowed_keys}

    existing_tags = metadata.get("Tags", [])
    update_payload["Tags"] = sorted(set(existing_tags + new_tags))

    if dry_run:
        logging.info(f"[DRY RUN] Would apply tags to {metadata.get('Name')} -> {update_payload['Tags']}")
        return

    try:
        client.jellyfin._post(f"Items/{item_id}", json=update_payload)
        logging.info(f"✅ Updated tags for {metadata.get('Name')}: {update_payload['Tags']}")
    except Exception as e:
        logging.error(f"❌ Failed to update tags: {e}")

# ----------------- MAIN SCRIPT ------------------
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--plex-url', required=True)
    parser.add_argument('--plex-token', required=True)
    parser.add_argument('--jellyfin-url', required=True)
    parser.add_argument('--jellyfin-username', required=True)
    parser.add_argument('--jellyfin-password', required=True)
    parser.add_argument('--match-by', choices=['tvdb', 'imdb'], default='tvdb')
    parser.add_argument('--dry-run', action='store_true')
    parser.add_argument('--log-to-stdout', action='store_true')
    args = parser.parse_args()

    setup_logging(to_stdout=args.log_to_stdout)

    client = jellyfin_connect(args.jellyfin_url, args.jellyfin_username, args.jellyfin_password)
    ids_to_labels = get_plex_items(args.plex_url, args.plex_token, args.match_by)

    total, tagged, skipped = len(ids_to_labels), 0, 0
    for external_id, data in ids_to_labels.items():
        title, labels = data['title'], data['labels']
        print(f"🔍 {title} ({args.match_by.upper()} ID {external_id})")
        item_id, name = find_jellyfin_item(client, args.match_by, external_id)
        if item_id:
            apply_tags(client, item_id, labels, args.dry_run)
            tagged += 1
        else:
            logging.warning(f"⚠️ No match in Jellyfin for {external_id} ({title})")
            skipped += 1

    print("\n=== SUMMARY ===")
    print(f"Matched Plex items: {total}")
    print(f"{'Would tag' if args.dry_run else 'Tagged'} in Jellyfin: {tagged}")
    print(f"Skipped or unmatched: {skipped}")

if __name__ == "__main__":
    main()
`

**Comment by luigi311** [comment]
Looking at this more this does seem out of scope. Perhaps stquinn script will fill your need for now
